ecs-agent/daemonimages/csidriver/driver/node.go

```diff
@@ -15,22 +15,224 @@ package driver
+	if volCap == nil {
+		return nil, status.Error(codes.InvalidArgument, "Volume capability not provided")
+	}
+
+	if !isValidVolumeCapabilities([]*csi.VolumeCapability{volCap}) {
+		return nil, status.Error(codes.InvalidArgument, "Volume capability not supported")
+	}
+	volumeContext := req.GetVolumeContext()
+	if isValidVolumeContext := isValidVolumeContext(volumeContext); !isValidVolumeContext {
+		return nil, status.Error(codes.InvalidArgument, "Volume Attribute is not valid")
+	}
+
+	// If the access type is block, do nothing for stage
+	switch volCap.GetAccessType().(type) {
+	case *csi.VolumeCapability_Block:
+		return &csi.NodeStageVolumeResponse{}, nil
+	}
+
+	mountVolume := volCap.GetMount()
+	if mountVolume == nil {
+		return nil, status.Error(codes.InvalidArgument, "NodeStageVolume: mount is nil within volume capability")
+	}
+
+	fsType := mountVolume.GetFsType()
+	if len(fsType) == 0 {
+		fsType = defaultFsType
+	}
+
+	_, ok := ValidFSTypes[strings.ToLower(fsType)]
+	if !ok {
+		return nil, status.Errorf(codes.InvalidArgument, "NodeStageVolume: invalid fstype %s", fsType)
+	}
+
+	context := req.GetVolumeContext()
+
+	blockSize, err := recheckParameter(context, BlockSizeKey, FileSystemConfigs, fsType)
+	if err != nil {
+		return nil, err
+	}
+	inodeSize, err := recheckParameter(context, INodeSizeKey, FileSystemConfigs, fsType)
+	if err != nil {
+		return nil, err
+	}
+	bytesPerINode, err := recheckParameter(context, BytesPerINodeKey, FileSystemConfigs, fsType)
+	if err != nil {
+		return nil, err
+	}
+	numINodes, err := recheckParameter(context, NumberOfINodesKey, FileSystemConfigs, fsType)
+	if err != nil {
+		return nil, err
+	}
+
+	mountOptions := collectMountOptions(fsType, mountVolume.MountFlags)
+
+	if ok = d.inFlight.Insert(volumeID); !ok {
+		return nil, status.Errorf(codes.Aborted, VolumeOperationAlreadyExists, volumeID)
+	}
+	defer func() {
+		klog.V(4).InfoS("NodeStageVolume: volume operation finished", "volumeID", volumeID)
+		d.inFlight.Delete(volumeID)
+	}()
+
+	devicePath, ok := req.PublishContext[DevicePathKey]
+	if !ok {
+		return nil, status.Error(codes.InvalidArgument, "Device path not provided")
+	}
+
+	partition := ""
+	if part, ok := volumeContext[VolumeAttributePartition]; ok {
+		if part != "0" {
+			partition = part
+		} else {
+			klog.InfoS("NodeStageVolume: invalid partition config, will ignore.", "partition", part)
+		}
+	}
+
+	source, err := d.findDevicePath(devicePath, volumeID, partition)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to find device path %s. %v", devicePath, err)
+	}
+
+	exists, err := d.mounter.PathExists(target)
+	klog.InfoS("NodeStageVolume: path exists:", "exists", exists)
+	if err != nil {
+		klog.InfoS("NodeStageVolume: path exists:", "err", err)
+		msg := fmt.Sprintf("failed to check if target %q exists: %v", target, err)
+		return nil, status.Error(codes.Internal, msg)
+	}
+	// When exists is true it means target path was created but device isn't mounted.
```

Comment:
When will this happen? Another volume had been mounted to this path and was unmounted?